### ISSUE TITLE
Subscribing for OUTLINE often doesn't result in data until next modification

### ISSUE THREAD

**Issue by DanTup** [dart-lang/sdk#30238]
Currently it's not possible to request outlines from the analysis server on-demand - there's #27064 about adding this, but at the time is wasn't really important as we had a workaround. This week I noticed some strange behaviour in Dart Code - the symbol list was working rather intermittently (it seemed to just hang quite a lot).

I didn't get chance to investigate, but today I was trying to implement something else that requires outlines and I discovered that subscribing for outlines does not immediately cause a notification with the current outlines (I presume it used to, since when the symbol list was implemented nobody spotted any issues with it) however it does send them immediately on the next buffer change.

Here's a (trimmed) log:

```text
[6:13:23 PM]: ==> {"id":"11","method":"analysis.setSubscriptions","params":{"subscriptions":{"OUTLINE":["/Users/danny/Dev/Testing/test_breakpoints/lib/main.dart"]}}}
[6:13:23 PM]: <== {"id":"11"}
[6:13:51 PM]: ==> {"id":"12","method":"analysis.updateContent","params":{"files":{"/Users/danny/Dev/Testing/test_breakpoints/lib/main.dart":{"type":"change","edits":[{"offset":304,"length":0,"replacement":" ","id":""}]}}}}
[6:13:51 PM]: ==> {"id":"13","method":"analysis.updateContent", // SNIP
[6:13:51 PM]: ==> {"id":"14","method":"completion.getSuggestions","params":{"file":"/Users/danny/Dev/Testing/test_breakpoints/lib/main.dart","offset":305}}
[6:13:51 PM]: <== {"id":"12","result":{}}
[6:13:51 PM]: <== {"event":"server.status","params":{"analysis":{"isAnalyzing":true}}}
[6:13:51 PM]: <== {"id":"13","result":{}}
[6:13:51 PM]: <== {"id":"14","result":{"id":"0"}}
[6:13:51 PM]: <== {"event":"analysis.errors","params":{"file":"/Users/danny/Dev/Testing/test_breakpoints/lib/main.dart","errors":[]}}
[6:13:51 PM]: <== {"event":"completion.results","params": // SNIP
[6:13:51 PM]: <== {"event":"analysis.outline","params":{"file":"/Users/danny/Dev/Testing/test_breakpoints/lib/main.dart", // SNIP
[6:13:51 PM]: <== {"event":"server.status","params":{"analysis":{"isAnalyzing":false}}}
```

Subscription is made at 6:13:23 but results don't appear until 6:13:51 immediately after an `updateContent` was sent. This wasn't fluke, it's totally* reproducible, I can leave it 10 minutes and then type a character and it'll come through.

\* It's actually intermittent, I don't know why but sometimes it seems to work, but mostly it does not - possible it's affected by timing with `setPriorityFiles`, I haven't dug into this yet as I'm not sure if considered a bug.

It is very possible that this behaviour is by design (it's not documented that outline will be immediately sent for a file when you subscribe), however without either #27064 or the data being sent reliably, it breaks our ability to do things like get the symbol list (or the new Flutter decorations I'm testing out). Given the choice, I'd rather get the on-demand option, since (ab)using subscriptions this way could introduce our own races (we may soon have multiple things trying to subscribe/unsubscribe for the notifications, and they could end up unsubscribing each other!), but I don't know the difference in implementing them.

There is another workaround I can try which involves subscribing to OUTLINE always, and keeping the data for any open files in a cache, and then providing access to it from there. I'm not against doing this but since nothing else currently works this way and our usage is generally "only once" for each file I thought I should ask about this first.

(FYI @bwilkerson - not sure if notifying you on analyzer service issues makes it easier for you to pick them up?)

**Comment by DanTup** [comment]
I found an easier workaround for now; just force an empty edit:

```ts
	let files: { [key: string]: as.ChangeContentOverlay } = {};
	files[editor.document.fileName] = {
		type: "change",
		edits: [{ offset: 0, length: 0, replacement: "", id: "" }]
	};
	this.analyzer.analysisUpdateContent({ files: files });
```

**Comment by DanTup** [comment]
While testing I found another drawback to using the notifications instead of a request/response here... With notifications I have no ability to tie the results to a "version" of the document, so if the user is typing quickly, I often get back data that is "old" and therefore has incorrect offsets. This results in bad things.

If I could send a request/response then I could easily ensure the data returned is for the "latest" version of the document and discard it if it's not (eg. if the editor content "version" is different to the one I had when I initiated the request).

TL;DR: #27064 would be significantly more useful than changing this behaviour.

**Comment by DanTup** [comment]
(Sorry for the spam!)

I did a refactor - I now just subscribe to outline/highlights/occurrences for all priority files. When I need the data, I send a dummy edit (insert an empty string at position 0) which seems to force an update. This somewhat turns this into request/response and eliminates all the races we had where different functions with subscribe/unsubscribe each other (by calling `setSubscriptions`). It means more traffic over the wire, but it seems to work reliably (actually, I think I can change it to just be the current active file and reduce this further).

So, I think for now I'm good. If you believe this is by design I'm happy for you to close it. It's slightly wonky that we're turning three notification functions into request/response but as long as you'll keep accepting a dummy edit to force a re-send of this data, I don't think it's causing any issues.

**Comment by bwilkerson** [comment]
It isn't by design. When a client subscribes it ought to get a notification within a reasonable period of time.

@scheglov Looking at AnalysisServer.setAnalysisSubscriptions, it looks like it's using AnalysisDriver.getResult to try to force a new result to be computed so that the result will be sent across the result stream and the notification from server will happen. I suspect that if the cached result of analysis is valid that we're no longer producing any event on the stream and hence the notification never happens. Does that seem reasonable to you, and if so, any suggestions for how to fix it? (I suppose it's better to have multiple identical notifications than it is to miss one.

**Comment by scheglov** [comment]
@bwilkerson Yes, your analysis is correct. When the file is a priority file, we cache its analysis result.  And `AnalysisDriver.getResult()` does not send the cached result to the `results` stream.

However it would be expensive to send it every time, because navigation through the same file causes assists and fixes requests, which call `getResult()`. We don't want recompute and resend navigation every time when the caret moves.

So, we need another way to fix it.
I will write a test and add a fix.

**Comment by scheglov** [comment]
https://codereview.chromium.org/2989633002

**Comment by DanTup** [comment]
Currently I'm sending a dummy edit to force the update. It would be good to remove this, but if I do it always, people on older SDKs will get broken behaviour (hangs in the symbol list, etc.). An ideal solution would be that I can use the analysis server version and only send the dummy edit for "older" versions.

Is it reasonable to bump the analyzer server version for this? Or, has the recent bump from @bwilkerson not shipped yet and I could just use that version number?

**Comment by scheglov** [comment]
The fix has been [landed](https://github.com/dart-lang/sdk/commit/939d8aaf6dd4a21cb5d0721cec3b90562b3c5b14).

I leave it to @bwilkerson to decide about the version, I'm not sure what was changed in this area recently.

**Comment by bwilkerson** [comment]
The current version number was included as part of 1.25.0-dev.5.0, 1.25.0-dev.6.0, 1.25.0-dev.7.0, and 1.25.0-dev.8.0. If your users only use stable versions of the SDK (something I don't know how to determine), then using the current number would be sufficient. It's safer if we update the version number.

**Comment by bwilkerson** [comment]
Done: https://codereview.chromium.org/2985003002/

**Comment by DanTup** [comment]
Great, thanks! 👍
